exercises/bhuvan_bezawada/ex02_oo_basics/src/main.cpp

```diff
@@ -0,0 +1,257 @@
+
+  // Calculates and returns the area
+  double area() override { return (M_PI * radius * radius); }
+};
+
+/**
+ *  Triangle class
+ **/
+class Triangle : public Shape {
+
+// Own height and base
+private:
+  double height;
+  double base;
+
+// Method implementations
+public:
+  // Constructor
+  Triangle(double heightLen, double baseLen) : Shape("Triangle", 3) {
+    height = heightLen;
+    base = baseLen;
+  }
+
+  // Calculates and returns the perimeter
+  double perimeter() {
+    return (base + 2 * sqrt(height * height + (base * base) / 4));
+  }
+
+  // Calculates and returns the area
+  double area() { return (0.5 * base * height); }
+};
+
+/**
+ *  ShapeSorter class
+ **/
+class ShapeSorter {
+
+  /**
+   * Comparators for sort()
+   *
+   * http://www.cplusplus.com/reference/algorithm/sort/
+   **/
+  static bool compareArea(Shape *shape1, Shape *shape2) {
+    return shape1->area() > shape2->area();
+  }
+
+  static bool comparePerimeter(Shape *shape1, Shape *shape2) {
+    return shape1->perimeter() > shape2->perimeter();
+  }
+
+public:
+  // Method to print matching given type
+  void printMatchingType(vector<Shape *> shapes, string name) {
+    for (vector<Shape *>::const_iterator it = shapes.begin();
+         it != shapes.end(); ++it) {
+      if (((*it)->getShapeType()).compare(name) == 0) {
+        cout << (*it)->getShapeType() << endl;
+      }
+    }
+  }
+
+  // Method to print shapes of chosen number of sides
+  void printShapesMatchingNumSides(vector<Shape *> shapes, int numSides) {
+    for (vector<Shape *>::const_iterator it = shapes.begin();
+         it != shapes.end(); ++it) {
+      if (((*it)->getNumSides()) == numSides) {
+        cout << (*it)->getShapeType() << endl;
+      }
+    }
+  }
+
+  // Print the shapes by area descending
+  void printAreaDescending(vector<Shape *> shapes) {
+    vector<Shape *> sortedAreas = shapes;
+    sort(sortedAreas.begin(), sortedAreas.end(), ShapeSorter::compareArea);
+
+    for (vector<Shape *>::iterator it = sortedAreas.begin();
+         it != sortedAreas.end(); ++it) {
+      cout << (*it)->getShapeType() << "\t\t" << (*it)->area() << endl;
+    }
+  }
+
+  // Print the shapes by perimeter descending
+  void printPerimeterDescending(vector<Shape *> shapes) {
+    vector<Shape *> sortedPerimeters = shapes;
+    sort(sortedPerimeters.begin(), sortedPerimeters.end(),
+         ShapeSorter::comparePerimeter);
+
+    for (vector<Shape *>::iterator it = sortedPerimeters.begin();
+         it != sortedPerimeters.end(); ++it) {
+      cout << (*it)->getShapeType() << "\t\t" << (*it)->perimeter() << endl;
+    }
+  }
+};
+
+/**
+ *  Main method
+ **/
+int main() {
+  // Shapes
+  Square square(5);
+  Rectangle rectangle(5, 10);
+  Circle circle(5);
+  Triangle triangle(10, 5);
+
+  // Add shapes into a vector
+  vector<Shape *> shapes;
+  shapes.push_back(&square);
+  shapes.push_back(&rectangle);
+  shapes.push_back(&circle);
+  shapes.push_back(&triangle);
```

Comment:
Interesting, you have no memory leak here so technically what you have done is fine, but the vector `shapes` does not own the shapes, and that isn't terribly clear. Maybe these should be stored as const pointers in the vector? Otherwise you could have stored them as `std::unique_ptr`